metric_tank/cluster.go

```diff
@@ -0,0 +1,213 @@
+	return &MetricPersistHandler{
+		metrics: metrics,
+	}
+}
+
+func (h *MetricPersistHandler) HandleMessage(m *nsq.Message) error {
+	ms := PersistMessage{}
+	err := json.Unmarshal(m.Body, &ms)
+	if err != nil {
+		log.Error(3, "skipping message. %s", err)
+		return nil
+	}
+	if ms.Instance == *instance {
+		log.Debug("skipping message we generated. %s - %s:%d", ms.Instance, ms.Key, ms.T0)
+		return nil
+	}
+
+	// get metric
+	if agg, ok := metrics.Get(ms.Key); ok {
+		// get chunk
+		agg.(*AggMetric).SaveChunkByT0(ms.T0)
+	}
+
+	return nil
+}
+
+func InitCluster(instance string, initialState bool, metrics Metrics, stats met.Backend) {
+	clusterStatus = NewClusterStatus(instance, initialState)
+
+	// init producers
+	pCfg := nsq.NewConfig()
+	pCfg.UserAgent = "metrics_tank"
+	err := app.ParseOpts(pCfg, *producerOpts)
+	if err != nil {
+		log.Fatal(4, "failed to parse nsq producer options. %s", err)
+	}
+	hostPool = hostpool.NewEpsilonGreedy(strings.Split(*nsqdTCPAddrs, ","), 0, &hostpool.LinearEpsilonValueCalculator{})
+	producers = make(map[string]*nsq.Producer)
+
+	for _, addr := range strings.Split(*nsqdTCPAddrs, ",") {
+		producer, err := nsq.NewProducer(addr, pCfg)
+		if err != nil {
+			log.Fatal(4, "failed creating producer %s", err.Error())
+		}
+		producers[addr] = producer
+	}
+
+	// init consumers
+	cfg := nsq.NewConfig()
+	cfg.UserAgent = "metrics_tank_cluster"
+	err = app.ParseOpts(cfg, *consumerOpts)
+	if err != nil {
+		log.Fatal(4, "failed to parse nsq consumer options. %s", err)
+	}
+	cfg.MaxInFlight = *maxInFlight
+
+	consumer, err := insq.NewConsumer(*topicNotifyPersist, fmt.Sprintf("metric_tank_%06d#ephemeral", rand.Int()%999999), cfg, "metric_persist.%s", stats)
+	if err != nil {
+		log.Fatal(4, "Failed to create NSQ consumer. %s", err)
+	}
+	handler := NewMetricPersistHandler(metrics)
+	consumer.AddHandler(handler)
+
+	nsqdAdds := strings.Split(*nsqdTCPAddrs, ",")
+	if len(nsqdAdds) == 1 && nsqdAdds[0] == "" {
+		nsqdAdds = []string{}
+	}
+	err = consumer.ConnectToNSQDs(nsqdAdds)
+	if err != nil {
+		log.Fatal(4, "failed to connect to NSQDs. %s", err)
+	}
+	log.Info("persist consumer connected to nsqd")
+
+	lookupdAdds := strings.Split(*lookupdHTTPAddrs, ",")
+	if len(lookupdAdds) == 1 && lookupdAdds[0] == "" {
+		lookupdAdds = []string{}
+	}
+	err = consumer.ConnectToNSQLookupds(lookupdAdds)
+	if err != nil {
+		log.Fatal(4, "failed to connect to NSQLookupds. %s", err)
+	}
+}
+
+// Handlers for HTTP interface.
+// Handle requests for /cluster. POST to set primary flag, GET to get current state.
+func clusterStatusHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method == "GET" {
+		getClusterStatus(w, req)
+		return
+	}
+	if req.Method == "POST" {
+		setClusterStatus(w, req)
+		return
+	}
+	http.Error(w, "not found.", http.StatusNotFound)
+}
+
+func setClusterStatus(w http.ResponseWriter, req *http.Request) {
+	status := make(map[string]bool)
+	decoder := json.NewDecoder(req.Body)
```

Comment:
why json? doesn't that make it harder to do simple curl requests on the commandline ? i think the http api should be as simple as possible. i personally was thinking of a pure-GET url based api like:
```
/cluster <- get status
/cluster/setprimary <- set status to primary
/cluster/setsecondary <- set status to secondary
```

simpler to query with curl, and no parsing needed in go. url is enough to know what needs to happen.